zebra-chain/src/value_balance.rs

```diff
@@ -333,55 +287,72 @@ impl ValueBalance<NonNegative> {
     }
 
     /// To byte array
-    pub fn to_bytes(self) -> [u8; 32] {
-        let transparent = self.transparent.to_bytes();
-        let sprout = self.sprout.to_bytes();
-        let sapling = self.sapling.to_bytes();
-        let orchard = self.orchard.to_bytes();
-        match [transparent, sprout, sapling, orchard].concat().try_into() {
+    pub fn to_bytes(self) -> [u8; 40] {
+        match [
+            self.transparent.to_bytes(),
+            self.sprout.to_bytes(),
+            self.sapling.to_bytes(),
+            self.orchard.to_bytes(),
+            self.deferred.to_bytes(),
+        ]
+        .concat()
+        .try_into()
+        {
             Ok(bytes) => bytes,
             _ => unreachable!(
-                "Four [u8; 8] should always concat with no error into a single [u8; 32]"
+                "five [u8; 8] should always concat with no error into a single [u8; 40]"
             ),
         }
     }
 
     /// From byte array
     #[allow(clippy::unwrap_in_result)]
-    pub fn from_bytes(bytes: [u8; 32]) -> Result<ValueBalance<NonNegative>, ValueBalanceError> {
+    pub fn from_bytes(bytes: &[u8]) -> Result<ValueBalance<NonNegative>, ValueBalanceError> {
```

Comment:
This method will now panic if the slice is too short, could we keep the argument typed as an array but update the array's length to 40 so it panics in the `FromDisk` impl instead?

Reply:
I did it this way so we can deserialize both 32 and 40 bytes, which makes it backward-compatible, see #8732. I'm happy to go down a different path, though.

Reply:
> so we can deserialize both 32 and 40 bytes

I was referring to this code: https://github.com/zcashfoundation/zebra/blob/9c0c7ffc89a44b0d2a19e42031775dd63a4fc3c5/zebra-chain/src/value_balance.rs#L340-L344

Reply:
To clarify, the implementation in #8732 is merely a trimmed-down version of the one in this PR.